### PR TITLE
WIP: Rework parallel transform scan pattern (sycl backend)

### PR DESCRIPTION
**Pull request by reble**
Adding future<T> implementation to dpcpp backend. 
Rework parallel transform scan pattern: add async behavior.

### CODE REVIEW

**Comment by timmiesmith** [comment]
Please run clang-format to allow checks to progress.  The changes look good to me.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h [hunk 1]
```diff
@@ -366,8 +366,8 @@ __parallel_transform_reduce(_ExecutionPolicy&& __exec, _Up __u, _Cp __combine, _
 //------------------------------------------------------------------------
 template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _BinaryOperation, typename _InitType,
           typename _LocalScan, typename _GroupScan, typename _GlobalScan>
-oneapi::dpl::__internal::__enable_if_device_execution_policy<
-    _ExecutionPolicy, ::std::pair<oneapi::dpl::__internal::__difference_t<_Range2>, typename _InitType::__value_type>>
+oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy,
+                                                             __future<typename _InitType::__value_type>>
```
Comment (capatober) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:370]: Could you please add fully qualified namespace for `__future` to make sure we use correct namespace?

**Review by capatober — COMMENTED** [review]
Looks good overall. This is what we discussed previously

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 2]
```diff
@@ -398,9 +398,28 @@ class __future_base
     operator sycl::event() const { return __my_event; }
 };
 
-template <typename T>
+template <typename _T>
 class __future : public __future_base
 {
+    ::std::size_t __offset;
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:404]: I think that field should be commented.
And probably renamed, 'result_idx' for example or something like that.. it should "tell" about meaning...

**Comment by reble** [comment]
New PR has been created. #201
Will close this one, continue discussion there.

**Pull request closed without merging**